File: src/main/kotlin/samlang/ast/Position.kt
package samlang.ast

data class Position(val line: Int, val column: Int) : Comparable<Position> {

    override fun compareTo(other: Position): Int {
        val c = line.compareTo(other = other.line)
        return if (c != 0) c else column.compareTo(other = other.column)
    }

    override fun toString(): String = "$line:$column"

}

File: typescript-rewrite/samlang-core/src/ast/common/Position.ts
export default class Position {
  static readonly DUMMY: Position = new Position(-1, -1);

  constructor(public readonly line: number, public readonly column: number) {}

  readonly compareTo = (other: Position): number => {
    const c = this.line - other.line;
    return c !== 0 ? c : this.column - other.column;
  };

  readonly toString = (): string => `${this.line + 1}:${this.column + 1}`;
}
